modules/ml/src/main/java/org/apache/ignite/ml/inference/plugin/MLInferencePluginProvider.java

```diff
@@ -0,0 +1,195 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.ml.inference.plugin;
+
+import java.io.Serializable;
+import java.util.UUID;
+import org.apache.ignite.Ignite;
+import org.apache.ignite.IgniteLogger;
+import org.apache.ignite.cache.CacheAtomicityMode;
+import org.apache.ignite.cache.CacheMode;
+import org.apache.ignite.cluster.ClusterNode;
+import org.apache.ignite.configuration.CacheConfiguration;
+import org.apache.ignite.configuration.IgniteConfiguration;
+import org.apache.ignite.ml.inference.storage.descriptor.ModelDescriptorStorageFactory;
+import org.apache.ignite.ml.inference.storage.model.ModelStorageFactory;
+import org.apache.ignite.plugin.CachePluginContext;
+import org.apache.ignite.plugin.CachePluginProvider;
+import org.apache.ignite.plugin.ExtensionRegistry;
+import org.apache.ignite.plugin.IgnitePlugin;
+import org.apache.ignite.plugin.PluginConfiguration;
+import org.apache.ignite.plugin.PluginContext;
+import org.apache.ignite.plugin.PluginProvider;
+import org.apache.ignite.plugin.PluginValidationException;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * Machine learning inference plugin provider.
+ */
+public class MLInferencePluginProvider implements PluginProvider<MLInferencePluginConfiguration> {
+    /** Plugin name. */
+    private static final String ML_INFERENCE_PLUGIN_NAME = "ml-inference-plugin";
+
+    /** Plugin version/ */
+    private static final String ML_INFERENCE_PLUGIN_VERSION = "1.0.0";
+
+    /** Default number of model storage backups. */
+    private static final int MODEL_STORAGE_DEFAULT_BACKUPS = 1;
+
+    /** Default number of model descriptor storage backups. */
+    private static final int MODEL_DESCRIPTOR_STORAGE_DEFAULT_BACKUPS = 1;
+
+    /** Plugin configuration. */
+    private MLInferencePluginConfiguration cfg;
+
+    /** Ignite instance. */
+    private Ignite ignite;
+
+    /** Ignite logger. */
+    private IgniteLogger log;
+
+    /** {@inheritDoc} */
+    @Override public String name() {
+        return ML_INFERENCE_PLUGIN_NAME;
+    }
+
+    /** {@inheritDoc} */
+    @Override public String version() {
+        return ML_INFERENCE_PLUGIN_VERSION;
+    }
+
+    /** {@inheritDoc} */
+    @Override public String copyright() {
+        return null;
+    }
+
+    /** {@inheritDoc} */
+    @SuppressWarnings("unchecked")
+    @Override public <T extends IgnitePlugin> T plugin() {
+        return (T)new MLInferencePlugin();
+    }
+
+    /** {@inheritDoc} */
+    @Override public void initExtensions(PluginContext ctx, ExtensionRegistry registry) {
+        IgniteConfiguration igniteCfg = ctx.igniteConfiguration();
+
+        this.ignite = ctx.grid();
+        this.log = ctx.log(this.getClass());
+
+        if (igniteCfg.getPluginConfigurations() != null) {
```

Comment:
We could:
1. Ask user to provide configuration that contains information about our caches manually, but it looks a bit too verbose and error prone.
2. Use the same approach Ignite uses to start SYSTEM_CACHE, but in this case we will have to add ML dependency (or at least some ML code) into Core module. I would prefer to avoid it.
3. Start required caches from static initializer of some ML class so that if ML module in classpath the caches will be started. But in this case we need additional synchronization with Ignite start process and can't pass configuration parameters properly.
4. Use [Apache Ignite Plugins](https://apacheignite.readme.io/docs/plugins). This approach looks good to me because it provides a way to pass configuration, provides entry point to run our code (syncronized with Ignite start) and this approach described in official Ignite docs.